sql/driver/operators_it_test.go

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ */
+
+package driver_test
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	hz "github.com/hazelcast/hazelcast-go-client"
+	"github.com/hazelcast/hazelcast-go-client/internal/it"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/sql/driver"
+)
+
+type employee struct {
+	ID            int `json:"id"`
+	PersonnelInfo `json:"personnel_info"`
+	CompanyInfo   `json:"company_info"`
+}
+type PersonnelInfo struct {
+	Name string `json:"name"`
+	Age  int    `json:"age"`
+}
+type CompanyInfo struct {
+	Title   string `json:"title"`
+	DepCode int    `json:"dep_code"`
+}
+
+func TestNestedJSONQuery(t *testing.T) {
+	it.SkipIf(t, "hz < 5.1")
+	it.SQLTester(t, func(t *testing.T, client *hz.Client, config *hz.Config, m *hz.Map, mapName string) {
+		defer driver.SetSerializationConfig(nil)
+		const rowCount = 50
+		db, err := initDB(t, config)
+		assert.Nil(t, err)
+		defer db.Close()
+		ms := createMappingStr(mapName, "bigint", "json")
+		it.Must(createMapping(t, db, ms))
+		// Insert random employee entries.
+		employees, err := populateMapWithEmployees(m, rowCount)
+		assert.Nil(t, err)
+		// Test query on nested json
+		complexQuery := fmt.Sprintf(`SELECT JSON_VALUE(this, '$.personnel_info.name') FROM "%s" WHERE CAST(JSON_VALUE(this, '$.id') AS TINYINT)=0`, mapName)
+		rows, err := db.QueryContext(context.Background(), complexQuery)
+		assert.Nil(t, err)
+		defer rows.Close()
+		var count int
+		for ; rows.Next(); count++ {
+			var name string
+			it.Must(rows.Scan(&name))
+			assert.Equal(t, employees[0].PersonnelInfo.Name, name)
+		}
+		assert.Equal(t, 1, count)
+	})
+}
+
+func TestJSONOperators(t *testing.T) {
+	it.SkipIf(t, "hz < 5.1")
+	it.SQLTester(t, func(t *testing.T, client *hz.Client, config *hz.Config, m *hz.Map, mapName string) {
+		const rowCount = 50
+		defer driver.SetSerializationConfig(nil)
+		db, err := initDB(t, config)
+		assert.Nil(t, err)
+		defer db.Close()
+		ms := createMappingStr(mapName, "bigint", "json")
+		it.Must(createMapping(t, db, ms))
+		// Create a view to test.
+		createViewQuery := fmt.Sprintf(`CREATE VIEW Personnel AS SELECT CAST(JSON_VALUE(this, '$.id') AS int) as pid, JSON_QUERY(this, '$.personnel_info') as info FROM "%s"`, mapName)
+		it.MustValue(db.Exec(createViewQuery))
+		// Insert random employee entries.
+		employees, err := populateMapWithEmployees(m, rowCount)
+		assert.Nil(t, err)
+		ctx := context.Background()
+		rows, err := db.QueryContext(ctx, `SELECT pid, info FROM "Personnel" ORDER BY pid ASC`)
+		assert.Nil(t, err)
+		defer rows.Close()
+		for i := 0; rows.Next(); i++ {
+			var pid int
+			var data serialization.JSON
+			it.Must(rows.Scan(&pid, &data))
+			assert.Equal(t, employees[i].ID, pid)
+			infoStr := it.MustValue(json.Marshal(employees[i].PersonnelInfo))
+			assert.Equal(t, serialization.JSON(infoStr.([]byte)), data)
```

Comment:
How about moving the type cast to the first line, like:
```
info := it.MustValue(json.Marshal(employees[i].PersonnelInfo)).([]byte
```

Reply:
Thanks, fixed at https://github.com/hazelcast/hazelcast-go-client/pull/726/commits/972309f6841a6c712338f131d67636935ede4025.